File: long_questions/DSA/Leetcode/golang/20_Trie/208_Implement_Trie_Prefix_Tree.go
package main

import "fmt"

// TrieNode represents a node in the trie
type TrieNode struct {
	children map[rune]*TrieNode
	isEnd    bool
}

// Trie represents the trie data structure
type Trie struct {
	root *TrieNode
}

// Constructor creates a new Trie
func Constructor() Trie {
	return Trie{
		root: &TrieNode{
			children: make(map[rune]*TrieNode),
			isEnd:    false,
		},
	}
}

// Insert inserts a word into the trie
// Time: O(L), Space: O(L) where L is word length
func (this *Trie) Insert(word string) {
	node := this.root
	
	for _, char := range word {
		if _, exists := node.children[char]; !exists {
			node.children[char] = &TrieNode{
				children: make(map[rune]*TrieNode),
				isEnd:    false,
			}
		}
		node = node.children[char]
	}
	
	node.isEnd = true
}

// Search returns true if the word is in the trie
// Time: O(L), Space: O(1)
func (this *Trie) Search(word string) bool {
	node := this.root
	
	for _, char := range word {
		if _, exists := node.children[char]; !exists {
			return false
		}
		node = node.children[char]
	}
	
	return node.isEnd
}

// StartsWith returns true if there is any word in the trie that starts with the given prefix
// Time: O(L), Space: O(1)
func (this *Trie) StartsWith(prefix string) bool {
	node := this.root
	
	for _, char := range prefix {
		if _, exists := node.children[char]; !exists {
			return false
		}
		node = node.children[char]
	}
	
	return true
}

// Helper function to visualize the trie structure
func (this *Trie) Visualize() {
	fmt.Println("Trie Structure:")
	this.visualizeHelper(this.root, "", true)
}

func (this *Trie) visualizeHelper(node *TrieNode, prefix string, isLast bool) {
	if node == nil {
		return
	}
	
	// Print current node
	if node.isEnd {
		fmt.Printf("%s└── (END)\n", prefix)
	} else {
		fmt.Printf("%s└── (node)\n", prefix)
	}
	
	// Print children
	count := 0
	for char, child := range node.children {
		count++
		newPrefix := prefix
		if isLast {
			newPrefix += "    "
		} else {
			newPrefix += "│   "
		}
		
		fmt.Printf("%s└── %c\n", newPrefix, char)
		this.visualizeHelper(child, newPrefix+"    ", count == len(node.children))
	}
}

// Additional useful methods

// GetAllWords returns all words in the trie
func (this *Trie) GetAllWords() []string {
	var words []string
	this.getAllWordsHelper(this.root, "", &words)
	return words
}

func (this *Trie) getAllWordsHelper(node *TrieNode, current string, words *[]string) {
	if node.isEnd {
		*words = append(*words, current)
	}
	
	for char, child := range node.children {
		this.getAllWordsHelper(child, current+string(char), words)
	}
}

// CountWords returns the number of words in the trie
func (this *Trie) CountWords() int {
	return this.countWordsHelper(this.root)
}

func (this *Trie) countWordsHelper(node *TrieNode) int {
	count := 0
	if node.isEnd {
		count++
	}
	
	for _, child := range node.children {
		count += this.countWordsHelper(child)
	}
	
	return count
}

// Delete removes a word from the trie
func (this *Trie) Delete(word string) bool {
	return this.deleteHelper(this.root, word, 0)
}

func (this *Trie) deleteHelper(node *TrieNode, word string, depth int) bool {
	if depth == len(word) {
		if !node.isEnd {
			return false // Word doesn't exist
		}
		node.isEnd = false
		return len(node.children) == 0
	}
	
	char := rune(word[depth])
	child, exists := node.children[char]
	if !exists {
		return false // Word doesn't exist
	}
	
	shouldDeleteChild := this.deleteHelper(child, word, depth+1)
	if shouldDeleteChild {
		delete(node.children, char)
		return len(node.children) == 0 && !node.isEnd
	}
	
	return false
}

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Trie (Prefix Tree) for String Operations
- **Tree Structure**: Each node represents a prefix, children for next characters
- **Prefix Sharing**: Common prefixes share nodes, saving space
- **Character Navigation**: Navigate tree character by character
- **End Markers**: Boolean flags mark word boundaries

## 2. PROBLEM CHARACTERISTICS
- **String Operations**: Insert, search, prefix matching
- **Prefix Optimization**: Share common prefixes efficiently
- **Dynamic Structure**: Support insertions and deletions
- **Alphabet Support**: Handle character sets (ASCII, Unicode)

## 3. SIMILAR PROBLEMS
- Design Add and Search Words (LeetCode 211) - Trie with wildcards
- Word Search II (LeetCode 212) - 2D grid with Trie
- Implement Magic Dictionary (LeetCode 676) - Trie for word transformations
- Concatenated Words (LeetCode 472) - Trie for word combinations

## 4. KEY OBSERVATIONS
- **Prefix Sharing**: Common prefixes share path in tree
- **Space Efficiency**: Shared prefixes reduce memory usage
- **Time Complexity**: O(L) for operations where L is string length
- **Node Structure**: Children map + end marker

## 5. VARIATIONS & EXTENSIONS
- **Compressed Trie**: Radix tree for space optimization
- **Ternary Search Trie**: Three-way branching for search optimization
- **Patricia Trie**: Path compression for long strings
- **Different Alphabets**: Support Unicode, case sensitivity

## 6. INTERVIEW INSIGHTS
- Always clarify: "What character set? Case sensitivity? Deletion needed?"
- Edge cases: empty strings, single characters, duplicate words
- Time complexity: O(L) for operations, O(total characters) space
- Key insight: prefix sharing saves space and enables prefix operations
- Memory vs Time tradeoff: More children vs compression

## 7. COMMON MISTAKES
- Not handling empty strings correctly
- Wrong character type (rune vs byte vs string)
- Not marking end of words properly
- Memory leaks in node creation
- Incorrect deletion logic

## 8. OPTIMIZATION STRATEGIES
- **Standard Trie**: O(L) time, O(total characters) space
- **Array Children**: Use fixed-size array for known alphabets
- **Compressed Trie**: Path compression for space efficiency
- **Lazy Deletion**: Mark nodes as deleted instead of immediate removal

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like a dictionary organized by prefixes:**
- You have a tree where each path represents a word prefix
- Common prefixes share the same path (like "app" and "apple")
- Each node knows if it's the end of a complete word
- Navigation follows characters down the tree
- Like a phone's T9 predictive text system

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Collection of words for prefix tree
2. **Goal**: Efficient insert, search, and prefix operations
3. **Constraint**: Optimize for prefix-based operations
4. **Output**: Trie data structure with required operations

#### Phase 2: Key Insight Recognition
- **"Prefix sharing natural fit"** → Common prefixes share nodes
- **"Tree structure"** → Hierarchical character organization
- **"End markers"** → Need to distinguish prefixes from complete words
- **"Character navigation"** → Follow tree path character by character

#### Phase 3: Strategy Development
```
Human thought process:
"I need a data structure for efficient prefix operations.
A trie is perfect:
- Each node represents a prefix
- Children represent next possible characters
- Common prefixes share the same path
- Navigation follows characters down the tree

Operations:
- Insert: Follow/create path for word
- Search: Navigate path, check end marker
- StartsWith: Navigate path, don't need end marker"
```

#### Phase 4: Edge Case Handling
- **Empty string**: Handle as valid word or ignore based on requirements
- **Single character**: Create immediate child of root
- **Duplicate words**: Handle gracefully (already exists)
- **Character set**: Support required alphabet (ASCII, Unicode)

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Words: ["apple", "app", "application"]

Human thinking:
"I'll build a trie step by step:

Insert "apple":
- root -> 'a' -> 'p' -> 'p' -> 'l' -> 'e' (END)
- Path: root-a-p-p-l-e, mark 'e' as word end

Insert "app":
- root -> 'a' -> 'p' -> 'p' (END)
- Path: root-a-p-p, mark 'p' as word end
- Note: "app" prefix is shared with "apple"

Insert "application":
- root -> 'a' -> 'p' -> 'p' -> 'l' -> 'i' -> 'c' -> 'a' -> 't' -> 'i' -> 'o' -> 'n' (END)
- Path: root-a-p-p-l-i-c-a-t-i-o-n, mark 'n' as word end
- Note: "app" prefix is shared, "appl" prefix is shared

Final structure has shared prefixes ✓"
```

#### Phase 6: Intuition Validation
- **Why trie works**: Shared prefixes reduce memory and enable prefix operations
- **Why O(L)**: Each operation follows path of length L
- **Why space efficient**: Common prefixes share nodes
- **Why tree structure**: Natural hierarchical organization

### Common Human Pitfalls & How to Avoid Them
1. **"Why not use hash set?"** → No prefix operations, more memory
2. **"Should I use array for children?"** → Faster for known alphabets
3. **"What about case sensitivity?"** → Clarify character handling requirements
4. **"Can I optimize space?"** → Use compressed trie or array children
5. **"What about Unicode?"** → Use rune instead of byte for Go

### Real-World Analogy
**Like a phone's T9 predictive text system:**
- Each key press narrows down possible words
- Common prefixes are shared (like "app" for "apple", "app", "application")
- The system remembers word endings (end markers)
- As you type, it follows the character path
- Can suggest completions based on current prefix

### Human-Readable Pseudocode
```
class Trie:
    root = TrieNode with children map and isEnd flag
    
    function insert(word):
        node = root
        for char in word:
            if char not in node.children:
                node.children[char] = new TrieNode()
            node = node.children[char]
        node.isEnd = true
    
    function search(word):
        node = root
        for char in word:
            if char not in node.children:
                return false
            node = node.children[char]
        return node.isEnd
    
    function startsWith(prefix):
        node = root
        for char in prefix:
            if char not in node.children:
                return false
            node = node.children[char]
        return true
```

### Execution Visualization

### Example: Words = ["apple", "app", "application"]
```
Trie Construction:
Step 1: Insert "apple"
root
 └── 'a'
     └── 'p'
         └── 'p'
             └── 'l'
                 └── 'e' (END)

Step 2: Insert "app" (shares prefix)
root
 └── 'a'
     └── 'p'
         └── 'p' (END) ← New word end marker
             └── 'l'
                 └── 'e' (END)

Step 3: Insert "application" (shares longer prefix)
root
 └── 'a'
     └── 'p'
         └── 'p' (END)
             └── 'l'
                 └── 'e' (END)
                     └── 'i'
                         └── 'c'
                             └── 'a'
                                 └── 't'
                                     └── 'i'
                                         └── 'o'
                                             └── 'n' (END)

Final trie with shared prefixes ✓
```

### Key Visualization Points:
- **Prefix Sharing**: Common prefixes share the same path
- **End Markers**: Distinguish complete words from prefixes
- **Character Navigation**: Follow tree path character by character
- **Space Efficiency**: Shared prefixes reduce memory usage

### Memory Layout Visualization:
```
Trie Node Structure:
TrieNode {
    children: map[rune]*TrieNode
    isEnd: bool
}

Example node for 'p' in "app", "apple":
children = {
    'l': TrieNode{...},  // for "apple"
    'p': TrieNode{isEnd: true}  // for "app"
}
isEnd = false  // "app" ends here, "apple" continues
```

### Time Complexity Breakdown:
- **Insert**: O(L) time, O(L) space where L is word length
- **Search**: O(L) time, O(1) space
- **StartsWith**: O(P) time, O(1) space where P is prefix length
- **Space**: O(total characters) for all inserted words
- **Optimal**: Best for prefix-based operations

### Alternative Approaches:

#### 1. Hash Set (O(L) time, O(N*L) space)
```go
type HashSet struct {
    words map[string]bool
}

func (hs *HashSet) Insert(word string) {
    hs.words[word] = true
}

func (hs *HashSet) Search(word string) bool {
    return hs.words[word]
}

func (hs *HashSet) StartsWith(prefix string) bool {
    for word := range hs.words {
        if strings.HasPrefix(word, prefix) {
            return true
        }
    }
    return false
}
```
- **Pros**: Simple to implement
- **Cons**: O(N) for prefix operations, more memory

#### 2. Sorted Array (O(log N) time, O(N*L) space)
```go
type SortedArray struct {
    words []string
}

func (sa *SortedArray) Insert(word string) {
    // Insert while maintaining sorted order
    // Binary search for insertion point
}

func (sa *SortedArray) Search(word string) bool {
    // Binary search for exact match
}

func (sa *SortedArray) StartsWith(prefix string) bool {
    // Binary search for first word with prefix
    // Check range of words with same prefix
}
```
- **Pros**: Memory efficient for search
- **Cons**: Complex prefix operations, O(log N) insertions

#### 3. Array-Based Trie (O(1) children access, O(Σ) space)
```go
type ArrayTrieNode struct {
    children [26]*ArrayTrieNode  // For lowercase English
    isEnd    bool
}

type ArrayTrie struct {
    root *ArrayTrieNode
}
```
- **Pros**: O(1) child access, no map overhead
- **Cons**: Fixed alphabet size, memory inefficient for sparse usage

### Extensions for Interviews:
- **Compressed Trie**: Radix tree for space optimization
- **Ternary Search Trie**: Three-way branching for search optimization
- **Patricia Trie**: Path compression for long strings
- **Different Alphabets**: Support Unicode, case sensitivity
- **Performance Analysis**: Discuss worst-case scenarios
*/
func main() {
	// Test cases
	trie := Constructor()
	
	// Test insertion and search
	fmt.Println("=== Testing Insert and Search ===")
	words := []string{"apple", "app", "application", "apt", "bat", "batch"}
	for _, word := range words {
		trie.Insert(word)
		fmt.Printf("Inserted: %s\n", word)
	}
	
	// Test search
	searchWords := []string{"apple", "app", "appl", "bat", "batter"}
	for _, word := range searchWords {
		found := trie.Search(word)
		fmt.Printf("Search %s: %t\n", word, found)
	}
	
	// Test starts with
	fmt.Println("\n=== Testing StartsWith ===")
	prefixes := []string{"app", "ap", "bat", "cat"}
	for _, prefix := range prefixes {
		found := trie.StartsWith(prefix)
		fmt.Printf("StartsWith %s: %t\n", prefix, found)
	}
	
	// Test additional methods
	fmt.Println("\n=== Additional Methods ===")
	fmt.Printf("Total words: %d\n", trie.CountWords())
	
	allWords := trie.GetAllWords()
	fmt.Printf("All words: %v\n", allWords)
	
	// Test deletion
	fmt.Println("\n=== Testing Delete ===")
	trie.Delete("app")
	fmt.Printf("After deleting 'app', Search 'app': %t\n", trie.Search("app"))
	fmt.Printf("After deleting 'app', Search 'apple': %t\n", trie.Search("apple"))
	fmt.Printf("Total words after deletion: %d\n", trie.CountWords())
	
	// Test edge cases
	fmt.Println("\n=== Testing Edge Cases ===")
	emptyTrie := Constructor()
	fmt.Printf("Empty trie search 'test': %t\n", emptyTrie.Search("test"))
	fmt.Printf("Empty trie startsWith 'test': %t\n", emptyTrie.StartsWith("test"))
	
	emptyTrie.Insert("")
	fmt.Printf("Insert empty string, count: %d\n", emptyTrie.CountWords())
}

File: long_questions/DSA/Leetcode/java/20_Trie/208_Implement_Trie_Prefix_Tree.java
import java.util.*;

public class ImplementTriePrefixTree {
    
    // TrieNode represents a node in trie
    public static class TrieNode {
        Map<Character, TrieNode> children;
        boolean isEnd;
        
        public TrieNode() {
            children = new HashMap<>();
            isEnd = false;
        }
    }

    // Trie represents trie data structure
    public static class Trie {
        private TrieNode root;

        // Constructor creates a new Trie
        public Trie() {
            root = new TrieNode();
        }

        // Insert inserts a word into trie
        // Time: O(L), Space: O(L) where L is word length
        public void insert(String word) {
            TrieNode node = root;
            
            for (char c : word.toCharArray()) {
                if (!node.children.containsKey(c)) {
                    node.children.put(c, new TrieNode());
                }
                node = node.children.get(c);
            }
            
            node.isEnd = true;
        }

        // Search returns true if word is in trie
        // Time: O(L), Space: O(1)
        public boolean search(String word) {
            TrieNode node = root;
            
            for (char c : word.toCharArray()) {
                if (!node.children.containsKey(c)) {
                    return false;
                }
                node = node.children.get(c);
            }
            
            return node.isEnd;
        }

        // StartsWith returns true if there is any word in trie that starts with given prefix
        // Time: O(L), Space: O(1)
        public boolean startsWith(String prefix) {
            TrieNode node = root;
            
            for (char c : prefix.toCharArray()) {
                if (!node.children.containsKey(c)) {
                    return false;
                }
                node = node.children.get(c);
            }
            
            return true;
        }

        // Helper function to visualize trie structure
        public void visualize() {
            System.out.println("Trie Structure:");
            visualizeHelper(root, "", true);
        }

        private void visualizeHelper(TrieNode node, String prefix, boolean isLast) {
            if (node == null) {
                return;
            }
            
            // Print children
            int count = 0;
            for (Map.Entry<Character, TrieNode> entry : node.children.entrySet()) {
                count++;
                char c = entry.getKey();
                TrieNode child = entry.getValue();
                
                String newPrefix = prefix;
                if (isLast) {
                    newPrefix += "    ";
                } else {
                    newPrefix += "│   ";
                }
                
                System.out.printf("%s└── %c\n", newPrefix, c);
                visualizeHelper(child, newPrefix + "    ", count == node.children.size());
            }
            
            // Print current node
            if (node.isEnd) {
                System.out.printf("%s└── (END)\n", prefix);
            }
        }

        // Additional useful methods

        // GetAllWords returns all words in trie
        public List<String> getAllWords() {
            List<String> words = new ArrayList<>();
            getAllWordsHelper(root, "", words);
            return words;
        }

        private void getAllWordsHelper(TrieNode node, String current, List<String> words) {
            if (node.isEnd) {
                words.add(current);
            }
            
            for (Map.Entry<Character, TrieNode> entry : node.children.entrySet()) {
                getAllWordsHelper(entry.getValue(), current + entry.getKey(), words);
            }
        }

        // CountWords returns number of words in trie
        public int countWords() {
            return countWordsHelper(root);
        }

        private int countWordsHelper(TrieNode node) {
            int count = 0;
            if (node.isEnd) {
                count++;
            }
            
            for (TrieNode child : node.children.values()) {
                count += countWordsHelper(child);
            }
            
            return count;
        }

        // Delete removes a word from trie
        public boolean delete(String word) {
            return deleteHelper(root, word, 0);
        }

        private boolean deleteHelper(TrieNode node, String word, int depth) {
            if (depth == word.length()) {
                if (!node.isEnd) {
                    return false; // Word doesn't exist
                }
                node.isEnd = false;
                return node.children.size() == 0;
            }
            
            char c = word.charAt(depth);
            TrieNode child = node.children.get(c);
            if (child == null) {
                return false; // Word doesn't exist
            }
            
            boolean shouldDeleteChild = deleteHelper(child, word, depth + 1);
            if (shouldDeleteChild) {
                node.children.remove(c);
                return node.children.size() == 0 && !node.isEnd;
            }
            
            return false;
        }
    }

    public static void main(String[] args) {
        // Test cases
        Trie trie = new Trie();
        
        // Test insertion and search
        System.out.println("=== Testing Insert and Search ===");
        String[] words = {"apple", "app", "application", "apt", "bat", "batch"};
        for (String word : words) {
            trie.insert(word);
            System.out.printf("Inserted: %s\n", word);
        }
        
        // Test search
        String[] searchWords = {"apple", "app", "appl", "bat", "batter"};
        for (String word : searchWords) {
            boolean found = trie.search(word);
            System.out.printf("Search %s: %b\n", word, found);
        }
        
        // Test starts with
        System.out.println("\n=== Testing StartsWith ===");
        String[] prefixes = {"app", "ap", "bat", "cat"};
        for (String prefix : prefixes) {
            boolean found = trie.startsWith(prefix);
            System.out.printf("StartsWith %s: %b\n", prefix, found);
        }
        
        // Test additional methods
        System.out.println("\n=== Additional Methods ===");
        System.out.printf("Total words: %d\n", trie.countWords());
        
        List<String> allWords = trie.getAllWords();
        System.out.printf("All words: %s\n", allWords);
        
        // Test deletion
        System.out.println("\n=== Testing Delete ===");
        trie.delete("app");
        System.out.printf("After deleting 'app', Search 'app': %b\n", trie.search("app"));
        System.out.printf("After deleting 'app', Search 'apple': %b\n", trie.search("apple"));
        System.out.printf("Total words after deletion: %d\n", trie.countWords());
        
        // Test edge cases
        System.out.println("\n=== Testing Edge Cases ===");
        Trie emptyTrie = new Trie();
        System.out.printf("Empty trie search 'test': %b\n", emptyTrie.search("test"));
        System.out.printf("Empty trie startsWith 'test': %b\n", emptyTrie.startsWith("test"));
        
        emptyTrie.insert("");
        System.out.printf("Insert empty string, count: %d\n", emptyTrie.countWords());
    }

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Trie (Prefix Tree)
- **Trie Structure**: Tree-like data structure for efficient prefix operations
- **Character Nodes**: Each node contains children for next characters
- **Prefix Search**: Efficiently check if any word starts with prefix
- **Space Efficiency**: Shared prefixes save memory

## 2. PROBLEM CHARACTERISTICS
- **String Operations**: Insert, search, prefix matching
- **Character-based**: Each level represents one character
- **Prefix Tree**: Common prefixes are shared among words
- **Dynamic Operations**: Support insert, delete, search operations

## 3. SIMILAR PROBLEMS
- Implement Trie (Word Dictionary)
- Word Search II
- Design Add and Search Words
- Replace Words

## 4. KEY OBSERVATIONS
- Each node contains map of children characters
- isEnd flag marks complete words
- Prefix search only needs to traverse, not check isEnd
- Insert creates nodes as needed for new words
- Delete requires careful backtracking

## 5. VARIATIONS & EXTENSIONS
- Array-based children (26 letters for lowercase)
- Compressed trie (radix tree)
- Support for Unicode characters
- Different character sets (digits, uppercase)

## 6. INTERVIEW INSIGHTS
- Clarify: "What character set should I support?"
- Edge cases: empty string, duplicate words, single character
- Time complexity: O(L) vs O(N*L) naive search
- Space complexity: O(total characters) vs O(N*L) separate storage

## 7. COMMON MISTAKES
- Using array of size 26 instead of HashMap for sparse data
- Not handling isEnd flag correctly
- Memory leaks in delete operation
- Not considering null children

## 8. OPTIMIZATION STRATEGIES
- Use array for fixed character sets (faster)
- Use HashMap for sparse character sets (memory efficient)
- Lazy deletion (mark as deleted)
- Path compression for long common prefixes

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like organizing words in a dictionary:**
- You have words that share common prefixes
- Instead of storing each word separately, share prefixes
- Each level represents one character position
- Navigate down the tree to spell words
- Common prefixes are shared among multiple words

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Collection of words to store
2. **Goal**: Support insert, search, prefix operations
3. **Output**: Efficient prefix tree implementation

#### Phase 2: Key Insight Recognition
- **"How to share prefixes?"** → Tree structure where common prefixes are shared
- **"What operations needed?"** → Insert, search, prefix check, delete
- **"How to navigate?"** → Follow character by character path
- **"When does word end?"** → isEnd flag marks complete words

#### Phase 3: Strategy Development
```
Human thought process:
"I'll implement a trie:
1. Root node has no character, just children
2. Each node has:
   - Map of children (character → node)
   - isEnd flag (marks complete word)
3. Insert: traverse/create nodes for each character
4. Search: follow path, check isEnd at end
5. Prefix: follow path, success if path exists"
```

#### Phase 4: Edge Case Handling
- **Empty string**: Handle specially (no characters to traverse)
- **Single character**: Simple one-level tree
- **Duplicate words**: Share existing nodes
- **Non-existent words**: Return false gracefully

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Words: ["apple", "app", "application"]

Human thinking:
"Let's build the trie step by step:

Insert "apple":
- Root → 'a' → 'p' → 'p' → 'l' → 'e'
- Mark 'e' node asEnd = true

Insert "app":
- Root → 'a' → 'p' → 'p'
- 'p' node already exists, use it
- Mark 'p' node asEnd = true

Insert "application":
- Root → 'a' → 'p' → 'p' → 'l' → 'i' → 'c' → 'a' → 't' → 'i' → 'o' → 'n'
- Mark 'n' node asEnd = true

Search "app":
- Root → 'a' → 'p' → 'p'
- 'p' node has isEnd = true ✓

Search "appl":
- Root → 'a' → 'p' → 'p' → 'l'
- 'l' node exists but isEnd = false ✗

Prefix "app":
- Root → 'a' → 'p' → 'p'
- Path exists ✓ (don't need to check isEnd)"
```

#### Phase 6: Intuition Validation
- **Why it works**: Tree structure naturally shares common prefixes
- **Why it's efficient**: O(L) operations vs O(N*L) naive
- **Why it's correct**: Each word has unique path in tree

### Common Human Pitfalls & How to Avoid Them
1. **"Why not just use array?"** → O(N*L) search vs O(L) trie
2. **"What about HashMap?"** → Can't efficiently handle prefix search
3. **"How to handle delete?"** → Complex backtracking needed
4. **"What about memory?"** → Shared prefixes save space

### Real-World Analogy
**Like organizing contacts in a phone:**
- You have contacts (words) with shared prefixes
- Instead of storing each contact separately, organize by first letter
- Within each letter, organize by second letter, etc.
- When searching for "John", you quickly find all "J" contacts
- Common prefixes like "App" are shared among "Apple", "Application"
- This is exactly how phone contacts are organized

### Human-Readable Pseudocode
```
class Trie:
    root = new TrieNode()
    
    function insert(word):
        node = root
        for character in word:
            if character not in node.children:
                node.children[character] = new TrieNode()
            node = node.children[character]
        node.isEnd = true
    
    function search(word):
        node = root
        for character in word:
            if character not in node.children:
                return false
            node = node.children[character]
        return node.isEnd
    
    function startsWith(prefix):
        node = root
        for character in prefix:
            if character not in node.children:
                return false
            node = node.children[character]
        return true
```

### Execution Visualization

### Example: Words = ["apple", "app", "application"]
```
Trie Structure:
Root
└── 'a'
    └── 'p'
        ├── 'p' (END) → "app"
        └── 'l'
            └── 'e' (END) → "apple"
            └── 'i'
                └── 'c'
                    └── 'a'
                        └── 't'
                            └── 'i'
                                └── 'o'
                                    └── 'n' (END) → "application"

Operations:
Search "app": Root→'a'→'p'→'p' ✓ (END found)
Search "appl": Root→'a'→'p'→'p'→'l' ✗ (no END)
Prefix "app": Root→'a'→'p'→'p' ✓ (path exists)
```

### Key Visualization Points:
- **Tree structure** naturally shares common prefixes
- **isEnd flags** mark complete words
- **Character navigation** follows word paths
- **Prefix search** only needs path existence

### Memory Layout Visualization:
```
Node Structure:
TrieNode {
    children: Map<Character, TrieNode>
    isEnd: boolean
}

Insert "apple":
Root.children['a'] = new Node()
Node['a'].children['p'] = new Node()
Node['a']['p'].children['p'] = new Node()
Node['a']['p']['p'].children['l'] = new Node()
Node['a']['p']['p']['l'].children['e'] = new Node()
Node['a']['p']['p']['l']['e'].isEnd = true

Shared prefixes save memory:
"app" and "apple" share: Root→'a'→'p'→'p'
Only "apple" continues: →'l'→'e'
```

### Time Complexity Breakdown:
- **Insert**: O(L) where L is word length
- **Search**: O(L) where L is word length
- **Prefix**: O(P) where P is prefix length
- **Space**: O(total characters) for all words
- **Optimal**: Much better than O(N*L) naive approaches
*/
